<a href="https://colab.research.google.com/github/kendraveren/text-mining/blob/main/TextMining.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

TUGAS TEXT MINING

Kendra Veren Widya A - DS0213

Dataset diambil dari https://github.com/ialfina/id-hatespeech-detection

Pertama kita harus menginstall nltk

In [309]:
pip install nltk

Lalu menginstall Sastrawi.

Sastrawi digunakan untuk melakukan pemrosesan teks dalam Bahasa Indonesia

In [310]:
pip install Sastrawi

Selanjutnya menginstall modul yang diperlukan

In [311]:
import pandas as pd
import numpy as np
import nltk
import os
import nltk.corpus
import re
import matplotlib.pyplot as plt

In [312]:
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

In [313]:
factory = StemmerFactory()
stemmer = factory.create_stemmer()

In [314]:
from google.colab import files

uploaded = files.upload()

Saving LabelTweet.txt to LabelTweet (7).txt


In [315]:
f = open("LabelTweet.txt", "r")
print(f.read())

Label	Tweet
Non_HS	RT @spardaxyz: Fadli Zon Minta Mendagri Segera Menonaktifkan Ahok Jadi Gubernur DKI https:\/\/t.co\/KH5vIRwPdO
Non_HS	RT @baguscondromowo: Mereka terus melukai aksi dalam rangka memenjarakan Ahok atau Ahok gagal dalam Pilkada.
Non_HS	Sylvi: bagaimana gurbernur melakukan kekerasan perempuan? Buktinya banyak ibu2 mau foto bareng #DebatFinalPilkadaJKT
Non_HS	"Ahmad Dhani Tak Puas Debat Pilkada, Masalah Jalan di Bekasi Belum Sempat Terungkap https:\/\/t.co\/m3HopPLUID\u2026"
Non_HS	RT @lisdaulay28: Waspada KTP palsu.....kawal PILKADA https:\/\/t.co\/OOoERQV4SM
Non_HS	"@felixsiauw masih bnyak hal agamis yg bisa dibahas, kenapa ke pilkada terus, apa tingkat keimanan anda cuma sebatas pilkada????"
Non_HS	@saidaqil apa maksud pernyataan. Anda? Apa anda mengaminkan kriminalisasi Ahok?? Hati2 jg dg mulutmu pak! https:\/\/t.co\/H4ODVaY0Md
Non_HS	Ahok Akan Integrasikan Transportasi Publik dengan Sistem Single Ticketing\u2026 https:\/\/t.co\/42PH7pqV0b #infoTribun
Non_HS	"Ahok ma

In [316]:
test = pd.read_csv('LabelTweet.txt', sep='\t', header=0)

In [317]:
tweet_df = test

In [318]:
tweet_df.head()

Label                                              Tweet
0  Non_HS  RT @spardaxyz: Fadli Zon Minta Mendagri Segera...
1  Non_HS  RT @baguscondromowo: Mereka terus melukai aksi...
2  Non_HS  Sylvi: bagaimana gurbernur melakukan kekerasan...
3  Non_HS  Ahmad Dhani Tak Puas Debat Pilkada, Masalah Ja...
4  Non_HS  RT @lisdaulay28: Waspada KTP palsu.....kawal P...

In [319]:
df = pd.DataFrame(tweet_df[['Label', 'Tweet']])

In [320]:
df

Label                                              Tweet
0    Non_HS  RT @spardaxyz: Fadli Zon Minta Mendagri Segera...
1    Non_HS  RT @baguscondromowo: Mereka terus melukai aksi...
2    Non_HS  Sylvi: bagaimana gurbernur melakukan kekerasan...
3    Non_HS  Ahmad Dhani Tak Puas Debat Pilkada, Masalah Ja...
4    Non_HS  RT @lisdaulay28: Waspada KTP palsu.....kawal P...
..      ...                                                ...
708      HS  Muka Si BABi Ahok Tuh Yg Mirip SERBET Lantai.....
709      HS  Betul bang hancurkan merka bang, musnahkan chi...
710      HS  Sapa Yg bilang Ahok anti korupsi!?, klo grombo...
711      HS  Gw juga ngimpi SENTILIN BIJI BABI AHOK, pcetar...
712      HS  Mudah2an gw ketemu sama SI BABI IWAN BOPENG DI...

[713 rows x 2 columns]

In [322]:
def remove_user(text):
  text = text.replace('\\t'," ").replace('\\n'," ").replace('\\u'," ").replace('\\',"")
  text = text.encode('ascii', 'replace').decode('ascii')
  text = ' '.join(re.sub("([@#][A-Za-z0-9]+)|(\w+:\/\/\S+)"," ", text).split())
  return text.replace("http://", " ").replace("https://", " ")

df['Tweet'] = df['Tweet'].apply(remove_user)

In [323]:
df

Label                                              Tweet
0    Non_HS  RT : Fadli Zon Minta Mendagri Segera Menonakti...
1    Non_HS  RT : Mereka terus melukai aksi dalam rangka me...
2    Non_HS  Sylvi: bagaimana gurbernur melakukan kekerasan...
3    Non_HS  Ahmad Dhani Tak Puas Debat Pilkada, Masalah Ja...
4    Non_HS           RT : Waspada KTP palsu.....kawal PILKADA
..      ...                                                ...
708      HS  Muka Si BABi Ahok Tuh Yg Mirip SERBET Lantai.....
709      HS  Betul bang hancurkan merka bang, musnahkan chi...
710      HS  Sapa Yg bilang Ahok anti korupsi!?, klo grombo...
711      HS  Gw juga ngimpi SENTILIN BIJI BABI AHOK, pcetar...
712      HS  Mudah2an gw ketemu sama SI BABI IWAN BOPENG DI...

[713 rows x 2 columns]

In [324]:
def remove(tweet):
    #remove angka
    tweet = re.sub('[0-9]+', '', tweet)

    #remove stock market tickers Like $GE
    tweet = re.sub(r'\$\w*', '', tweet)

     #remove old style retweet text "RT"
    tweet = re.sub(r'^RT[\s]+', '', tweet)

    #remove hashtag
    #only removing the hash # sign from word
    tweet = re.sub(r'#', '', tweet)
    return tweet
df['Tweet'] = df['Tweet'].apply(remove)

In [325]:
df

Label                                              Tweet
0    Non_HS  : Fadli Zon Minta Mendagri Segera Menonaktifka...
1    Non_HS  : Mereka terus melukai aksi dalam rangka memen...
2    Non_HS  Sylvi: bagaimana gurbernur melakukan kekerasan...
3    Non_HS  Ahmad Dhani Tak Puas Debat Pilkada, Masalah Ja...
4    Non_HS              : Waspada KTP palsu.....kawal PILKADA
..      ...                                                ...
708      HS  Muka Si BABi Ahok Tuh Yg Mirip SERBET Lantai.....
709      HS  Betul bang hancurkan merka bang, musnahkan chi...
710      HS  Sapa Yg bilang Ahok anti korupsi!?, klo grombo...
711      HS  Gw juga ngimpi SENTILIN BIJI BABI AHOK, pcetar...
712      HS  Mudahan gw ketemu sama SI BABI IWAN BOPENG DI ...

[713 rows x 2 columns]

In [326]:
#menghapus duplikat
df.drop_duplicates(subset ="Tweet", keep = 'first', inplace = True)

Setelah menghapus duplikasi tweet, jumlah tweet yang awalnya 713 berkurang menjadi 702

In [327]:
df

Label                                              Tweet
0    Non_HS  : Fadli Zon Minta Mendagri Segera Menonaktifka...
1    Non_HS  : Mereka terus melukai aksi dalam rangka memen...
2    Non_HS  Sylvi: bagaimana gurbernur melakukan kekerasan...
3    Non_HS  Ahmad Dhani Tak Puas Debat Pilkada, Masalah Ja...
4    Non_HS              : Waspada KTP palsu.....kawal PILKADA
..      ...                                                ...
708      HS  Muka Si BABi Ahok Tuh Yg Mirip SERBET Lantai.....
709      HS  Betul bang hancurkan merka bang, musnahkan chi...
710      HS  Sapa Yg bilang Ahok anti korupsi!?, klo grombo...
711      HS  Gw juga ngimpi SENTILIN BIJI BABI AHOK, pcetar...
712      HS  Mudahan gw ketemu sama SI BABI IWAN BOPENG DI ...

[702 rows x 2 columns]

In [328]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [329]:
#import stopword
from nltk.corpus import stopwords
stopwords_indonesia = stopwords.words("indonesian")

#import sastrawi
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory
factory = StemmerFactory()
stemmer = factory.create_stemmer()

#Tokenize
from nltk.tokenize import TweetTokenizer

# Happy Emoticons
emoticons_happy = set([
    ':-)', ':)', ';)', ':o)', ':]', ':3', ':c)', ':>', '=]', '8)', '=)', ':}',
    ':^)', ':-D', ':D', '8-D', '8D', 'x-D', 'xD', 'X-D', 'XD', '=-D', '=D',
    '=-3', '=3', ':-))', ":'-)", ":')", ':*', ':^*', '>:P', ':-P', ':P', 'X-P',
    'x-p', 'xp', 'XP', ':-p', ':p', '=p', ':-b', ':b', '>:)', '>;)', '>:-)',
    '<3'
    ])
 
# Sad Emoticons
emoticons_sad = set([
    ':L', ':-/', '>:/', ':S', '>:[', ':@', ':-(', ':[', ':-||', '=L', ':<',
    ':-[', ':-<', '=\\', '=/', '>:(', ':(', '>.<', ":'-(", ":'(", ':\\', ':-c',
    ':c', ':{', '>:\\', ';('
    ])
 
# all emoticons (happy + sad)
emoticons = emoticons_happy.union(emoticons_sad)
 
def clean_tweets(tweet):
    # remove stock market tickers like $GE
    tweet = re.sub(r'\$\w*', '', tweet)
 
    # remove old style retweet text "RT"
    tweet = re.sub(r'^RT[\s]+', '', tweet)
 
    # remove hyperlinks
    tweet = re.sub(r'https?:\/\/.*[\r\n]*', '', tweet)
    
    # remove hashtags
    # only removing the hash # sign from the word
    tweet = re.sub(r'#', '', tweet)
    
    #remove coma
    tweet = re.sub(r',','',tweet)
    
    #remove angka
    tweet = re.sub('[0-9]+', '', tweet)
 
    # tokenize tweets
    tokenizer = TweetTokenizer(preserve_case=False, strip_handles=True, reduce_len=True)
    tweet_tokens = tokenizer.tokenize(tweet)
 
    tweets_clean = []    
    for word in tweet_tokens:
        if (word not in stopwords_indonesia and # remove stopwords
              word not in emoticons and # remove emoticons
                word not in string.punctuation): # remove punctuation
            #tweets_clean.append(word)
            stem_word = stemmer.stem(word) # stemming word
            tweets_clean.append(stem_word)
 
    return tweets_clean
df['Tweet'] = df['Tweet'].apply(clean_tweets)


In [330]:
df

Label                                              Tweet
0    Non_HS  [fadli, zon, mendagri, nonaktif, ahok, gubernu...
1    Non_HS  [luka, aksi, rangka, penjara, ahok, ahok, gaga...
2    Non_HS  [sylvi, gurbernur, keras, perempuan, bukti, fo...
3    Non_HS  [ahmad, dhani, puas, debat, pilkada, jalan, be...
4    Non_HS            [waspada, ktp, palsu, , kawal, pilkada]
..      ...                                                ...
708      HS  [muka, si, babi, ahok, tuh, yg, serbet, lantai...
709      HS  [bang, hancur, merka, bang, musnah, china, bab...
710      HS  [sapa, yg, bilang, ahok, anti, korupsi, klo, g...
711      HS  [gw, ngimpi, sentilin, biji, babi, ahok, pceta...
712      HS  [mudah, gw, ketemu, si, babi, iwan, bopeng, tp...

[702 rows x 2 columns]

In [331]:
#remove punct
def remove_punct(text):
    text  = " ".join([char for char in text if char not in string.punctuation])
    return text

In [332]:
df['Tweet'] = df['Tweet'].apply(remove_punct)

In [333]:
df

Label                                              Tweet
0    Non_HS      fadli zon mendagri nonaktif ahok gubernur dki
1    Non_HS   luka aksi rangka penjara ahok ahok gagal pilkada
2    Non_HS  sylvi gurbernur keras perempuan bukti foto bareng
3    Non_HS  ahmad dhani puas debat pilkada jalan bekas ungkap
4    Non_HS                    waspada ktp palsu kawal pilkada
..      ...                                                ...
708      HS         muka si babi ahok tuh yg serbet lantai btp
709      HS  bang hancur merka bang musnah china babi bumi ...
710      HS  sapa yg bilang ahok anti korupsi klo grombolan...
711      HS  gw ngimpi sentilin biji babi ahok pcetar pceta...
712      HS  mudah gw ketemu si babi iwan bopeng tps nya an...

[702 rows x 2 columns]

In [389]:
le1 = preprocessing.LabelEncoder()
df['Label'] =le1.fit_transform(df['Label'])

In [390]:
df

Label                                              Tweet
0        1      fadli zon mendagri nonaktif ahok gubernur dki
1        1   luka aksi rangka penjara ahok ahok gagal pilkada
2        1  sylvi gurbernur keras perempuan bukti foto bareng
3        1  ahmad dhani puas debat pilkada jalan bekas ungkap
4        1                    waspada ktp palsu kawal pilkada
..     ...                                                ...
708      0         muka si babi ahok tuh yg serbet lantai btp
709      0  bang hancur merka bang musnah china babi bumi ...
710      0  sapa yg bilang ahok anti korupsi klo grombolan...
711      0  gw ngimpi sentilin biji babi ahok pcetar pceta...
712      0  mudah gw ketemu si babi iwan bopeng tps nya an...

[702 rows x 2 columns]

In [391]:
tfidf_vectorizer = TfidfVectorizer(ngram_range=(1, 2),max_df=0.75, min_df=5, max_features=10000)

# TF-IDF feature matrix
tfidf = tfidf_vectorizer.fit_transform(df['Tweet'] )
tfidf

<702x283 sparse matrix of type '<class 'numpy.float64'>'
	with 3702 stored elements in Compressed Sparse Row format>

In [395]:
#Naive Bayes Modelling
X = tfidf
y = df['Label'].astype(int)
X_train_tfidf, X_test_tfidf, y_train, y_test = train_test_split(X.toarray(), y, random_state=42, test_size=0.30)
nbb=GaussianNB()
nbb.fit(X_train_tfidf,y_train)
y_preds = nbb.predict(X_test_tfidf)
acc2b=accuracy_score(y_test,y_preds)
report = classification_report( y_test, y_preds )
print(report)
print("Naive Bayes, Accuracy Score:",acc2b)

              precision    recall  f1-score   support

           0       0.61      0.75      0.67        71
           1       0.85      0.76      0.80       140

    accuracy                           0.75       211
   macro avg       0.73      0.75      0.74       211
weighted avg       0.77      0.75      0.76       211

Naive Bayes, Accuracy Score: 0.7535545023696683
